geode-core/src/main/java/org/apache/geode/internal/cache/PRHARedundancyProvider.java

```diff
@@ -1784,13 +1787,24 @@ public void run2() {
     // }
   }
 
-  void startRedundancyLogger(int proxyBuckets) {
-    if (proxyBuckets > 0) {
-      redundancyLogger = new RedundancyLogger(this);
-      Thread loggingThread = new LoggingThread(
-          "RedundancyLogger for region " + this.prRegion.getName(), false, this.redundancyLogger);
-      loggingThread.start();
-    }
+  void createRedundancyLoggerAndStartLoggingThread(int proxyBuckets) {
+    redundancyLogger = createRedundancyLogger();
+    allBucketsRecoveredFromDisk = createAllBucketsRecoveredFromDisk(proxyBuckets);
```

Comment:
I think we still have a problem. RedundancyLogger has this line in its constructor:
    this.allBucketsRecoveredFromDisk = redundancyProvider.getAllBucketsRecoveredFromDiskLatch();
With these changes I think this will cause this line to set allBucketsRecoveredFromDisk on the RedunancyLogger to null. Since we are passing "this" into the RedundancyLogger why don't you change the above line to "create" the AllBucketsRecoveredFromDiskLatch on the redundancyProvider instead of "getting" it. You would want to also pass in the proxyBucket cound to the RedundancyLogger so it could pass this on to the create method. And this would allow you to remove the "getAllBucketsRecoveredFromDiskLatch" method since the only caller is the above line. All the other code that uses the allBucketsRecoveredFromDisk on the provider already checks to make sure it is not null. And this would make sure that the allBucketsRecoveredFromDisk on the RedundancyLogger is never null.

Reply:
I cleaner solution, I think, is to change RedundancyLogger to PersistentBucketRecoverer. One of the things this class would do, it what it does now; log information on how persistent bucket recovery is doing. But that other thing it would do is own the allBucketsRecoveredFromDisk latch. The redundancyProvider would no longer have this latch. It already has an instance variable referring to the RedundancyLogger (you would need to change this to be volatile as the latch currently is). The methods that want to wait for recovery to complete will now just call a method on PersistentBucketRecoverer. Basically any work with this latch should be implemented on the PersistentBucketRecoverer. The provider will just check to see if its PersistentBucketRecoverer is not null and if it has one delegate to it.